csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll

```diff
@@ -0,0 +1,118 @@
+    this.getQualifiedName().matches("System.Data.DataSet")
+  }
+}
+
+/**
+ * A Class that include a property or generic collection of type `DataSet` and `DataTable`
+ */
+class ClassWithDataSetOrTableMember extends DataSetOrTableRelatedClass {
+  ClassWithDataSetOrTableMember() {
+    exists( Property p |
+      p = this.getAProperty() |
+      p.getType() instanceof DataSetOrTable
+    ) or exists ( AssignableMember am |
+      am = this.getAField() or
+      am = this.getAMember() |
+      am.getType() instanceof DataSetOrTable
+    ) or exists( Property p |
+      p = this.getAProperty() |
+      p.getType() instanceof DataSetOrTable or
+      p.getType().(ConstructedGeneric).getATypeArgument() instanceof DataSetOrTable 
+    )
+  }
+}
+
+/**
+ * Serializable types
+ */
+class SerializableClass extends Class {
+  SerializableClass() {
+    (
+      this.getABaseType*().getQualifiedName().matches("System.Xml.Serialization.XmlSerializer") or
+      this.getABaseInterface*().getQualifiedName().matches("System.Runtime.Serialization.ISerializable") or
+      this.getABaseType*().getQualifiedName().matches("System.Runtime.Serialization.XmlObjectSerializer") or
+      this.getABaseInterface*().getQualifiedName().matches("System.Runtime.Serialization.ISerializationSurrogateProvider") or
+      this.getABaseType*().getQualifiedName().matches("System.Runtime.Serialization.XmlSerializableServices") or
+      this.getABaseInterface*().getQualifiedName().matches("System.Xml.Serialization.IXmlSerializable")
+    ) or exists( Attribute a |
+      a = this.getAnAttribute() |
+      a.getType().getQualifiedName().toString() = "System.SerializableAttribute"
+    )
+  }
+}
+
+/**
+ * Holds if the serializable class `c` has a property or field `m` that is of `DataSet` or `DataTable` related type
+ */
+predicate isClassUnsafeXmlSerializerImplementation( SerializableClass c, Member m) {
+  exists( Property p |
+    m = p |
+    p = c.getAProperty() and
+    p.getType() instanceof DataSetOrTableRelatedClass
+  ) or exists ( AssignableMember am |
+    am = m |
+    ( am = c.getAField() or am = c.getAMember() ) and
+    am.getType() instanceof DataSetOrTableRelatedClass
+  )
+}
+
+/**
+ * Serializable class that has a property or field that is of `DataSet` or `DataTable` related type
+ */
+class UnsafeXmlSerializerImplementation extends SerializableClass {
+  UnsafeXmlSerializerImplementation() {
+    isClassUnsafeXmlSerializerImplementation( this, _ )
+  }
+}
+
+/**
+ * Method that may be unsafe when used to deserialize DataSet and DataTable related types
+ */
+class UnsafeXmlReadMethod extends Method {
+  UnsafeXmlReadMethod() {
+    this.getQualifiedName().toString() = "System.Data.DataTable.ReadXml" or
+    this.getQualifiedName().toString() = "System.Data.DataTable.ReadXmlSchema" or
+    this.getQualifiedName().toString() = "System.Data.DataSet.ReadXml" or
+    this.getQualifiedName().toString() = "System.Data.DataSet.ReadXmlSchema" or
```

Comment:
IMHO `ReadXmlSchema` sink will cause false positives.
A call to `ds.ReadXml(reader, XmlReadMode.IgnoreSchema)` made after `ds.ReadXmlSchema(reader);` makes it vulnerable.
A call to `ds.ReadXml(reader)` is always vulnerable.
A call to `ds.ReadXml(reader, XmlReadMode.IgnoreSchema)` with ds that has hardcoded schema is always safe.

Maybe splitting it to queries with different certainties would be an option?

Reply:
Only `cs/dataset-serialization/unsafe-type-used-data-contract-serializer` has been marked as precision = high. I can change it to medium. Would that be acceptable? Thanks

Reply:
I would rather split the query into two parts. The `ReadXmlSchema` query could be low precision and `ReadXml` - medium. @hvitved  what do you think?